### PR TITLE
check for incorrect kinds (for 'real', 'integer' etc.)

### LINKED ISSUE #3450: Unhandled exception on passing unsupported "kind" parameter
```bash
> cat a.f90
program main
    integer(3) :: i
end program

> ./src/bin/lfortran a.f90           
Internal Compiler Error: Unhandled exception
Traceback (most recent call last):
  File "/Users/gxyd/OpenSource/lfortran/src/bin/lfortran.cpp", line 2391
    LCompilers::print_stack_on_segfault();
  File "/Users/gxyd/OpenSource/lfortran/src/bin/lfortran.cpp", line 2355
    err = compile_to_object_file(arg_file, tmp_o, false,
  File "/Users/gxyd/OpenSource/lfortran/src/bin/lfortran.cpp", line 949
    res = fe.get_llvm3(*asr, lpm, diagnostics, infile);
  File "/Users/gxyd/OpenSource/lfortran/src/lfortran/fortran_evaluator.cpp", line 360
    compiler_options, run_fn, infile);
  File "/Users/gxyd/OpenSource/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 9599
    v.visit_asr((ASR::asr_t&)asr);
  File "../libasr/asr.h", line 5109
  File "../libasr/asr.h", line 5085
  File "../libasr/asr.h", line 5110
  File "../libasr/asr.h", line 4816
  File "/Users/gxyd/OpenSource/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 945
    visit_symbol(*item.second);
  File "../libasr/asr.h", line 5112
  File "../libasr/asr.h", line 4824
  File "/Users/gxyd/OpenSource/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 3059
    declare_vars(x);
  File "/Users/gxyd/OpenSource/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 3399
    type = llvm_utils->get_type_from_ttype_t(
  File "/Users/gxyd/OpenSource/lfortran/src/libasr/codegen/llvm_utils.cpp", line 1409
    llvm_type = getIntType(a_kind);
  File "/Users/gxyd/OpenSource/lfortran/src/libasr/codegen/llvm_utils.cpp", line 1624
    LCOMPILERS_ASSERT(false);
AssertFailed: false
```

### PR DESCRIPTION
**Pull request by gxyd**
Fixes #3450 

an incorrect `kind` could be in the following way:
* incorrect data-type (e.g. string, float instead of integer)
* unsupported type (e.g. kind=32 for integer)

### CODE REVIEW

File: src/lfortran/semantics/ast_common_visitor.h [hunk 1]
```diff
@@ -2872,14 +2877,22 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
                     ASRUtils::type_get_past_allocatable(type)));
             }
         } else if (sym_type->m_type == AST::decl_typeType::TypeLogical) {
-            type = ASRUtils::TYPE(ASR::make_Logical_t(al, loc, compiler_options.po.default_integer_kind));
```
Comment (gxyd) [inline src/lfortran/semantics/ast_common_visitor.h:2875]: @certik , all this while has it been intentional to pass `default_integer_kind` as the kind, regardless of what the user passed?

Comment (gxyd) [reply src/lfortran/semantics/ast_common_visitor.h:2875]: e.g. `logical(8) :: x` would automatically change it to `4`

Comment (Shaikh-Ubaid) [reply src/lfortran/semantics/ast_common_visitor.h:2875]: At the ASR level we support logicals of only kind `4`. Kinds for `logical/boolean` type as such would not matter.

File: tests/errors/kind_string_of_int.f90 [hunk 2]
```diff
@@ -0,0 +1,4 @@
+program string_kind_integer
```
Comment (Shaikh-Ubaid) [inline tests/errors/kind_string_of_int.f90:1]: Should this be the following thereby matching the filename?
```suggestion
program kind_string_of_int
```

File: src/lfortran/semantics/ast_common_visitor.h (hunk 1 above)
Comment (gxyd) [reply src/lfortran/semantics/ast_common_visitor.h:2875]: > I think kinds for logical/boolean type as such would not matter.

That's true.

I don't have much knowledge in this regards, but what I do know is for when Fortran program need to interact with other languages or systems where different representations of logical values, having the same representation might be a problem.

Currently in `main` for the below program:
```f90
program main
    logical(8) :: x
    x = .TRUE.
    print *, kind(x)
end program
```

the output is `4`. Ideally I would like to raise an error, but maybe we can raise a warning for now?

File: tests/errors/kind_string_of_int.f90 (hunk 2 above)
Comment (gxyd) [reply tests/errors/kind_string_of_int.f90]: I've resolved this now

File: src/lfortran/semantics/ast_common_visitor.h (hunk 1 above)
Comment (Shaikh-Ubaid) [reply src/lfortran/semantics/ast_common_visitor.h:2877]: I do not have any opinion on this at the moment. Maybe create an issue for it that the kinds returned by LFortran and GFortran do match and we can tackle it in a separate PR.

Comment (gxyd) [reply src/lfortran/semantics/ast_common_visitor.h:2877]: fair enough. I've put a comment and restored the 'kind' for logical to `default_integer_kind`

**Comment by gxyd** [comment]
This MR is ready for review, let me know if it needs any improvements.

File: src/libasr/asr_utils.h [hunk 3]
```diff
@@ -2773,23 +2774,40 @@ inline int extract_kind(ASR::expr_t* kind_expr, const Location& loc) {
...
+        // allow integer binary operator kinds (e.g. '1 + 7')
+        case ASR::exprType::IntegerBinOp:
+        // allow integer kinds (e.g. 4, 8 etc.)
+        case ASR::exprType::IntegerConstant: {
+            // we still need to ensure that values are constant
+            // e.g. 'integer :: a = 4; real(1*a) :: x' is an invalid kind,
+            // as 'a' // isn't a constant.
+            // ToDo: we should raise a better error, by "locating" just
+            // 'a' as well, instead of the whole '1*a'
+            int a_kind {};
```
Comment (Shaikh-Ubaid) [inline src/libasr/asr_utils.h:2793]: Initialize to `-1` instead?
```suggestion
            int a_kind = -1;
```

File: src/lfortran/semantics/ast_common_visitor.h [hunk 4]
```diff
@@ -2874,6 +2879,13 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
                     ASRUtils::type_get_past_allocatable(type)));
             }
         } else if (sym_type->m_type == AST::decl_typeType::TypeLogical) {
+            if (a_kind != 1 && a_kind != 2 && a_kind != 4 && a_kind != 8) {
+                throw SemanticError("Kind " + std::to_string(a_kind) + " is not supported for Logical",
+                                sym_type->m_kind->loc);
+            }
+            // currently we change the kind's of all logical's to
+            // 'default_integer_kind'. GFortran support's logical's of
+            // different kind's, we need to think about this
```
Comment (Shaikh-Ubaid) [inline src/lfortran/semantics/ast_common_visitor.h:2888]: .

**Review by Shaikh-Ubaid — APPROVED** [review]
I shared a suggestion above. Rest looks good to me. Great work! Thanks for this!

File: src/libasr/asr_utils.h (hunk 3 above)
Comment (gxyd) [reply src/libasr/asr_utils.h]: I've changed its initialisation to -1.

**Comment by gxyd** [comment]
Thank you, @Shaikh-Ubaid, for taking the time to review it. Please feel free to suggest even the smallest details for improvement in PR's.

**Pull request merged**